File: Medium/701_Insert_Into_a_Binary_Search_Tree/Solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
*/


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


class Solution {
public:
    TreeNode* insertIntoBST(TreeNode* root, int val) {
        TreeNode* newNode = new TreeNode(val);

        if (root == NULL) return newNode;

        TreeNode* curr = root;

        while (curr != NULL) {
            if (val < curr->val) {
                if (curr->left != NULL) {
                    curr =  curr->left;
                } else {
                    curr->left = newNode;
                    break;
                }
            } else {
                if (curr->right != NULL) {
                    curr =  curr->right;
                } else {
                    curr->right = newNode;
                    break;
                }
            }
        }
        
        return root;
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;
    

    return 0;
}

File: Medium/701_Insert_Into_a_Binary_Search_Tree/Solution.java


public class Solution {
    public TreeNode insertIntoBST(TreeNode root, int val) {
        TreeNode treeNode = new TreeNode(val);

        if (root == null) {
            return treeNode;
        }

        TreeNode curr = root;

        while (curr != null) {
            if (curr.val < val) {
                if (curr.right == null) {
                    curr.right = treeNode;
                    break;
                } else {
                    curr = curr.right;
                }
            } else {
                if (curr.left == null) {
                    curr.left = treeNode;
                    break;
                } else {
                    curr = curr.left;
                }
            }
        }

        return root;
    }

    public static void main(String[] args) {
        

    }
}


// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}